intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h

```diff
@@ -740,6 +743,430 @@ static inline JBLAS_CODE decompress_s4_s8(utils::int4x2* srcptr, int8_t* dstptr,
+
+    std::vector<Xbyak::Zmm> reg_srcs(row_pack), reg_tmps(row_pack);
+    const int ZIDX_TranSrc = 0;
+    const int ZIDX_TransTmp = row_pack;
+    for (int i = 0; i < row_pack; i++) reg_srcs[i] = Xbyak::Zmm(ZIDX_TranSrc + i);
+    for (int i = 0; i < row_pack; i++) reg_tmps[i] = Xbyak::Zmm(ZIDX_TransTmp + i);
+
+    xor_(reg_iterrow, reg_iterrow);
+    L(".rowloop");
+    xor_(reg_itercol, reg_itercol);
+    mov(reg_tmp2.cvt32(), ptr[parambase + OFFSET(row)]);
+    sub(reg_tmp2, reg_iterrow);
+    cmp(reg_tmp2, row_pack);
+    jb(".tailrowloop", T_NEAR);
+
+    L(".colloop");
+    mov(reg_tmp1, reg_itercol);
+    imul(reg_tmp1, reg_dststride);
+    lea(reg_tmp, ptr[reg_dstptr + reg_tmp1]);
+    lea(reg_tmp1, ptr[reg_srcptr + reg_itercol * src_bytes]);
+    for (int jj = 0; jj < n_tile; jj += ne_zmm) {
+      generate_Nbitsmask(mask_rd, reg_itercol, ptr[reg_colsize - jj], reg_tmp2, reg_tmp3, ne_zmm);
+      for (int ii = 0; ii < row_pack; ii++) {
+        const Xbyak::Xmm reg_srcs_ii = src_bytes_vmm == 64   ? Xbyak::Zmm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 32 ? Xbyak::Ymm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 16 ? Xbyak::Xmm(reg_srcs[ii].getIdx())
+                                                             : (assert(false), reg_srcs[ii]);
+        if (src_bytes == 1) {
+          vmovdqu8(reg_srcs_ii | mask_rd | T_z, ptr[reg_tmp1 + ii * reg_srcstride + jj * src_bytes]);
+        } else if (src_bytes == 2) {
+          vmovdqu16(reg_srcs_ii | mask_rd | T_z, ptr[reg_tmp1 + ii * reg_srcstride + jj * src_bytes]);
+        } else if (src_bytes == 4) {
+          vmovdqu32(reg_srcs_ii | mask_rd | T_z, ptr[reg_tmp1 + ii * reg_srcstride + jj * src_bytes]);
+        }
+      }
+      if (src_t == JblasF32 && dst_t == JblasBF16) {
+        vcvtne2ps2bf16(reg_tmps[0], reg_srcs[1], reg_srcs[0]);
+        vpermt2w(reg_tmps[0], vreg_idx0, reg_tmps[0]);
+        vmovups(ptr[reg_tmp + jj * row_pack * dst_bytes], reg_tmps[0]);
+      } else {
+        // interleave_2rows_4regs(reg_srcs.data(), reg_tmps.data());
+        assert(false);  // Not implemented
+      }
+    }
+    add(reg_itercol, n_tile);
+    cmp(reg_itercol.cvt32(), ptr[parambase + OFFSET(col)]);
+    jb(".colloop");
+    lea(reg_srcptr, ptr[reg_srcptr + row_pack * reg_srcstride]);
+    lea(reg_dstptr, ptr[reg_dstptr + row_pack * n_tile * dst_bytes]);
+
+    add(reg_iterrow, row_pack);
+    cmp(reg_iterrow.cvt32(), ptr[parambase + OFFSET(row)]);
+    jb(".rowloop");
+    jmp(".aftercolloop", T_NEAR);
+
+    L(".tailrowloop");
+    L(".tailcolloop");
+    mov(reg_tmp1, reg_itercol);
+    imul(reg_tmp1, reg_dststride);
+    lea(reg_tmp, ptr[reg_dstptr + reg_tmp1]);
+    lea(reg_tmp1, ptr[reg_srcptr + reg_itercol * src_bytes]);
+    for (int jj = 0; jj < n_tile; jj += ne_zmm) {
+      generate_Nbitsmask(mask_rd, reg_itercol, ptr[reg_colsize - jj], reg_tmp2, reg_tmp3, ne_zmm);
+      if (row_pack == 2) {
```

Comment:
constexpr

Reply:
I tried to avoid the use of template in code generators to potentially save some time for compilation. It should be harmless to add some runtime branches in the logic of JIT code generation. As a result, it is not a constexpr here.